File: app/src/main/java/com/codekk/ui/activity/RecommendSearchActivity.java
package com.codekk.ui.activity;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.Toolbar;
import android.text.Html;
import android.text.TextUtils;
import android.view.View;

import com.codekk.R;
import com.codekk.mvp.model.RecommendSearchModel;
import com.codekk.mvp.presenter.RecommendSearchPresenterImpl;
import com.codekk.mvp.view.ViewManager;
import com.common.widget.LoadMoreRecyclerView;
import com.common.widget.StatusLayout;
import com.xadapter.OnXBindListener;
import com.xadapter.adapter.XRecyclerViewAdapter;
import com.xadapter.holder.XViewHolder;

import butterknife.BindView;
import com.codekk.ui.base.BaseStatusActivity;
import com.codekk.utils.UIUtils;

/**
 * by y on 2017/5/20.
 */

public class RecommendSearchActivity extends BaseStatusActivity<RecommendSearchPresenterImpl> implements ViewManager.RecommendSearchView,
        OnXBindListener<RecommendSearchModel.RecommendArrayBean>,
        LoadMoreRecyclerView.LoadMoreListener, SwipeRefreshLayout.OnRefreshListener {


    private static final String TEXT_KEY = "text";
    private String text;
    private int page = 1;

    @BindView(R.id.refreshLayout)
    SwipeRefreshLayout mRefresh;
    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.recyclerView)
    LoadMoreRecyclerView mRecyclerView;

    private XRecyclerViewAdapter<RecommendSearchModel.RecommendArrayBean> mAdapter;

    public static void newInstance(@NonNull String text) {
        Bundle bundle = new Bundle();
        bundle.putString(TEXT_KEY, text);
        UIUtils.startActivity(RecommendSearchActivity.class, bundle);
    }

    @Override
    protected void initBundle() {
        super.initBundle();
        text = bundle.getString(TEXT_KEY);
    }

    @Override
    protected void initCreate(@NonNull Bundle savedInstanceState) {
        mToolbar.setTitle(text);
        setSupportActionBar(mToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
//        mRecyclerView.setLoadingListener(this);
        mAdapter = new XRecyclerViewAdapter<>();

        mRecyclerView.setAdapter(
                mAdapter
                        .setLayoutId(R.layout.item_recommend_list)
                        .setOnItemClickListener((view, position, info) -> UIUtils.openBrowser(info.getUrl()))
                        .onXBind(this)
        );

        mRefresh.setOnRefreshListener(this);
        mRefresh.post(this::onRefresh);
    }

    @Override
    protected void clickNetWork() {
        super.clickNetWork();
        if (!mRefresh.isRefreshing()) {
            onRefresh();
        }
    }

    @Override
    protected RecommendSearchPresenterImpl initPresenter() {
        return new RecommendSearchPresenterImpl(this);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_recommend_search;
    }

    public void showProgress() {
        mRefresh.setRefreshing(true);
    }

    @Override
    public void hideProgress() {
        mRefresh.setRefreshing(false);
    }

    @Override
    public void onRefresh() {
        mStatusView.setStatus(StatusLayout.SUCCESS);
        mPresenter.netWorkRequest(text, page = 1);
    }

    @Override
    public void onLoadMore() {
        if (mRefresh.isRefreshing()) {
            return;
        }
        mPresenter.netWorkRequest(text, page);
    }


    @Override
    public void netWorkSuccess(RecommendSearchModel recommendSearchModel) {
        if (mStatusView != null) {
            if (page == 1) {
                mAdapter.removeAll();
            }
            ++page;
            mAdapter.addAllData(recommendSearchModel.getRecommendArray());
            mStatusView.setStatus(StatusLayout.SUCCESS);
        }
    }

    @Override
    public void netWorkError(Throwable e) {
        if (mStatusView != null) {
            if (page == 1) {
                mAdapter.removeAll();
                mStatusView.setStatus(StatusLayout.ERROR);
            } else {
                UIUtils.snackBar(mStatusView, R.string.net_error);
            }
        }
    }

    @Override
    public void noMore() {
        if (mStatusView != null) {
            if (page == 1) {
                mStatusView.setStatus(StatusLayout.EMPTY);
            } else {
                UIUtils.snackBar(mStatusView, R.string.data_empty);
            }
        }
    }

    @Override
    public void onXBind(XViewHolder holder, int position, RecommendSearchModel.RecommendArrayBean recommendArrayBean) {
        if (!TextUtils.isEmpty(recommendArrayBean.getTitle())) {
            holder.setTextView(R.id.tv_recommend_title, recommendArrayBean.getTitle());
        }
        AppCompatTextView descView = holder.getView(R.id.tv_recommend_desc);
        descView.setVisibility(TextUtils.isEmpty(recommendArrayBean.getDesc()) ? View.GONE : View.VISIBLE);
        if (!TextUtils.isEmpty(recommendArrayBean.getDesc())) {
            descView.setText(Html.fromHtml(recommendArrayBean.getDesc()));
        }
    }
}

File: app/src/main/java/com/codekk/ui/activity/RecommendSearchActivity.kt
package com.codekk.ui.activity

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.text.parseAsHtml
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.codekk.R
import com.codekk.databinding.LayoutListBinding
import com.codekk.ext.*
import com.codekk.mvp.presenter.impl.RecommendPresenterImpl
import com.codekk.mvp.view.RecommendListView
import com.codekk.ui.base.BaseViewBindActivity
import com.codekk.ui.widget.LoadMoreRecyclerView
import com.xadapter.*
import com.xadapter.adapter.XAdapter
import com.xadapter.holder.XViewHolder
import org.jetbrains.anko.browse

/**
 * by y on 2017/5/20.
 */
class RecommendSearchActivity : BaseViewBindActivity<LayoutListBinding, RecommendPresenterImpl>(), RecommendListView, LoadMoreRecyclerView.LoadMoreListener, SwipeRefreshLayout.OnRefreshListener {

    companion object {
        const val TEXT_KEY = "text"
    }

    private val text by lazy { bundle?.getString(TEXT_KEY, "").orEmpty() }
    private val mAdapter by lazy { XAdapter<RecommendListBean>() }
    private var page = 1

    override fun initViewBind(): LayoutListBinding {
        return LayoutListBinding.inflate(layoutInflater)
    }

    override fun initPresenter(): RecommendPresenterImpl {
        return RecommendPresenterImpl(this)
    }

    override fun initCreate(savedInstanceState: Bundle?) {
        baseViewBind.toolbarRoot.toolbar.title = text
        setSupportActionBar(baseViewBind.toolbarRoot.toolbar)

        viewBind.recyclerView.setHasFixedSize(true)
        viewBind.recyclerView.layoutManager = LinearLayoutManager(this)

        viewBind.recyclerView.adapter = mAdapter
                .setItemLayoutId(R.layout.item_recommend_list)
                .setOnItemClickListener { _, _, info -> browse(info.url, true) }
                .setOnBind { holder, _, entity ->
                    onXBind(holder, entity)
                }

        viewBind.refreshLayout.setOnRefreshListener(this)
        viewBind.refreshLayout.post { this.onRefresh() }
    }

    override fun clickNetWork() {
        super.clickNetWork()
        if (!viewBind.refreshLayout.isRefreshing) {
            onRefresh()
        }
    }

    override fun showProgress() {
        viewBind.refreshLayout.isRefreshing = true
    }

    override fun hideProgress() {
        viewBind.refreshLayout.isRefreshing = false
    }

    override fun onRefresh() {
        baseViewBind.statusLayout.success()
        mPresenter.netWorkRequestSearch(text, page = 1)
    }

    override fun onLoadMore() {
        if (viewBind.refreshLayout.isRefreshing) {
            return
        }
        mPresenter.netWorkRequestSearch(text, page)
    }


    override fun netWorkSuccess(entity: RecommendListModel) {
        if (page == 1) {
            mAdapter.removeAll()
        }
        ++page
        mAdapter.addAll(entity.recommendList)
    }

    override fun netWorkError(throwable: Throwable) {
        if (page == 1) {
            baseViewBind.statusLayout.error()
            mAdapter.removeAll()
        } else {
            baseViewBind.statusLayout.snackBar(R.string.net_error)
        }
    }

    override fun noMore() {
        if (page == 1) {
            baseViewBind.statusLayout.empty()
            mAdapter.removeAll()
        } else {
            baseViewBind.statusLayout.snackBar(R.string.data_empty)
        }
    }

    private fun onXBind(holder: XViewHolder, recommendArrayBean: RecommendListBean) {
        if (!TextUtils.isEmpty(recommendArrayBean.title)) {
            holder.setText(R.id.tv_recommend_title, recommendArrayBean.title)
        }
        val descView = holder.findById<AppCompatTextView>(R.id.tv_recommend_desc)
        descView.visibility = if (TextUtils.isEmpty(recommendArrayBean.desc)) View.GONE else View.VISIBLE
        if (!TextUtils.isEmpty(recommendArrayBean.desc)) {
            descView.text = recommendArrayBean.desc.parseAsHtml()
        }
    }

}
